### PR TITLE
Add `MakeService::into_service` and `MakeService::as_service`

### PR DESCRIPTION
**Pull request by davidpdrsn**
Resolves https://github.com/tower-rs/tower/issues/253

Adds `MakeService::into_service` and `MakeService::as_service` which converts `MakeService`s into `Service`s. `into_service` consumes `self` and `as_service` borrows `self` mutably.

### CODE REVIEW

**Comment by davidpdrsn** [comment]
One thought I had was that it would be nice if `IntoService` and `AsService` were public types so users can name them in type aliases and such. Not sure where to expose them though. `tower::make::{IntoService, AsService}` doesn't feel right and if we go with `tower::make::make_service::{IntoService, AsService}` it feels like `MakeService` should be moved in there as well.

File: tower/src/make/make_service.rs [hunk 1]
```diff
@@ -65,3 +147,51 @@ where
         Service::call(self, target)
     }
 }
+
+#[derive(Debug, Clone)]
```
Comment (hawkw) [inline tower/src/make/make_service.rs:151]: Unfortunately, I think we should probably provide manual `Clone` and `Debug` impls for this, rather than deriving them. The derived implementations require _all_ type parameters of the type deriving `Clone` or `Debug` to implement those traits. However, we don't actually need the request type to implement `Clone` to clone an `IntoService` or implement `Debug` to format one — it's only necessary because of the `PhantomData`.

Therefore, I think we should provide an `impl<M: Clone, Request> Clone for IntoService<M, Request>` and an `impl<M: fmt::Debug, Request> fmt::Debug for IntoService<M, Request>`, rather than deriving them.

File: tower/src/make/make_service.rs [hunk 2]
```diff
@@ -65,3 +147,51 @@ where
...
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.make.poll_ready(cx)
+    }
+
+    fn call(&mut self, target: Target) -> Self::Future {
+        self.make.make_service(target)
+    }
+}
+
+#[derive(Debug)]
```
Comment (hawkw) [inline tower/src/make/make_service.rs:175]: similarly, this should probably not be derived IMO.

File: tower/src/make/make_service.rs [hunk 3]
```diff
@@ -65,3 +147,51 @@ where
         Service::call(self, target)
     }
 }
+
+#[derive(Debug, Clone)]
+pub struct IntoService<M, Request> {
```
Comment (hawkw) [inline tower/src/make/make_service.rs:152]: let's make sure to add doc comments for these structs as well. it's fine if they just say 
```suggestion
/// Service returned by [`MakeService::into_service`][into]. 
/// 
/// See the documentation on [`into_service`][into] for details.
///
/// [into]: MakeService::into_service
#[derive(Debug, Clone)]
pub struct IntoService<M, Request> {
```
or something

File: tower/src/make/make_service.rs [hunk 4]
```diff
@@ -65,3 +147,51 @@ where
...
+        self.make.poll_ready(cx)
+    }
+
+    fn call(&mut self, target: Target) -> Self::Future {
+        self.make.make_service(target)
+    }
+}
+
+#[derive(Debug)]
+pub struct AsService<'a, M, Request> {
```
Comment (hawkw) [inline tower/src/make/make_service.rs:176]: similarly, a quick doc comment would be good
```suggestion
/// Service returned by [`MakeService::as_service`][as]. 
/// 
/// See the documentation on [`as_service`][as] for details.
///
/// [as]: MakeService::as_service
#[derive(Debug)]
pub struct AsService<'a, M, Request> {
```

**Review by hawkw — CHANGES_REQUESTED** [review]

File: tower/src/make/make_service.rs (hunk 1 above)
Comment (davidpdrsn) [reply tower/src/make/make_service.rs]: Ah yes ofc. Fixed in 4b1dd1505be29

File: tower/src/make/make_service.rs (hunk 2 above)
Comment (davidpdrsn) [reply tower/src/make/make_service.rs]: 4b1dd1505be29 👍

**Comment by davidpdrsn** [comment]
Should we expose them as `MakeServiceIntoService` and `MakeServiceAsService`? Both a bit wordy 🤔

**Comment by hawkw** [comment]
> Should we expose them as `MakeServiceIntoService` and `MakeServiceAsService`? Both a bit wordy thinking

as long as they  live in the `make` module and aren't re-exported at the crate root, I think  `make::IntoService` and `make::AsService` are better.

File: tower/src/make/make_service.rs [hunk 5]
```diff
@@ -65,3 +148,93 @@ where
...
+    type Error = M::Error;
+    type Future = M::Future;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.make.poll_ready(cx)
+    }
+
+    fn call(&mut self, target: Target) -> Self::Future {
+        self.make.make_service(target)
+    }
```
Comment (hawkw) [inline tower/src/make/make_service.rs:200]: TIOLI: these could probably be inlined...compiler might be smart enough to figure that out now though...

**Review by hawkw — APPROVED** [review]
great, this looks good to me! thanks!

**Comment by davidpdrsn** [comment]
> > Should we expose them as `MakeServiceIntoService` and `MakeServiceAsService`? Both a bit wordy thinking
> 
> as long as they live in the `make` module and aren't re-exported at the crate root, I think `make::IntoService` and `make::AsService` are better.

I was thinking if we wanted to make the types public since the `make` module is currently private.

File: tower/src/make/make_service.rs (hunk 5 above)
Comment (davidpdrsn) [reply tower/src/make/make_service.rs]: 6a753cc4f3383 👍

**Comment by hawkw** [comment]
> I was thinking if we wanted to make the types public since the `make` module is currently private.

Hmm...let's just make the `make` module public, then. We should continue re-exporting `MakeService` from the root of the crate, though, and maybe add `#[doc(inline)]` on the re-export.

**Comment by davidpdrsn** [comment]
👌

File: tower/src/make/mod.rs [hunk 6]
```diff
@@ -1,7 +1,9 @@
 //! Trait aliases for Services that produce specific types of Responses.
 
 mod make_connection;
-mod make_service;
+
+pub mod make_service;
 
 pub use self::make_connection::MakeConnection;
+#[doc(inline)]
 pub use self::make_service::MakeService;
```
Comment (hawkw) [inline tower/src/make/mod.rs:9]: Sorry, I misunderstood what you were saying earlier. This is the `tower::make` module, which is public, and contains a private module named `make_service`. I thought you were saying the `make` module itself was not public.

Let's not make `tower::make::make_service` pub. Let's just re-export `AsService` and `IntoService` _in_ the `tower::make` module:
```suggestion
mod make_service;

pub use self::make_connection::MakeConnection;
pub use self::make_service::{MakeService, AsService, IntoService};
```

I think this is fine --- it's clear that these are `MakeService` adapters because they're in the `tower::make` module.

**Review by hawkw — CHANGES_REQUESTED** [review]

File: tower/src/make/mod.rs (hunk 6 above)
Comment (davidpdrsn) [reply tower/src/make/mod.rs]: Haha all good

**Review by hawkw — APPROVED** [review]
looks good to me!
